### PR TITLE
feat(frontend/backend/shared): add deadline for tasks bb-361

### PR DESCRIPTION
**Pull request by luiqor**
#361 
#### Backend:
- [x] TaskService `create` method:
  - Designed to be used in AI service to create tasks
  - Request DTO: `categoryId`, `description`, `label`, `user` (UserDto)

- [x] `calculateDeadline`:
  - Calculates the deadline as the next closest day of the week to the current day.
  - Before calculation, Sunday must be normalized. A 1-7 system is used across the BeBalance app, where Sunday is 7. JavaScript `Date.getDay`, `Date.setDate` use a 0-6 system, where Sunday is 0.

#### Frontend:
- [x] `Deadline` component
- [x] Countdown logic
- [x] Expired task graphics update logic

#### Preview
The countdown updates every minute. The update can be seen at ~1:00 in the preview.

https://github.com/user-attachments/assets/962adc5e-0a4c-42f8-abd3-538fba6a658f

### CODE REVIEW

File: apps/backend/src/modules/tasks/task.service.ts [hunk 1]
```diff
@@ -1,26 +1,76 @@
...
 import { type TaskRepository } from "./task.repository.js";
 
 class TaskService implements Service {
+	private calculateDeadline = (userTaskDays: number[]): string => {
+		const createdAt = new Date();
+		const createdAtDayOfWeek = createdAt.getDay();
+
+		const normalizeTaskDays = userTaskDays.map((day) =>
+			day === Sunday.USER_TASK ? Sunday.NORMALIZED : day,
+		);
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.service.ts:27]: B27

File: apps/backend/src/modules/tasks/task.service.ts [hunk 2]
```diff
@@ -1,26 +1,76 @@
...
 import { TaskEntity } from "./task.entity.js";
 import { type TaskModel } from "./task.model.js";
 import { type TaskRepository } from "./task.repository.js";
 
 class TaskService implements Service {
+	private calculateDeadline = (userTaskDays: number[]): string => {
+		const createdAt = new Date();
+		const createdAtDayOfWeek = createdAt.getDay();
+
+		const normalizeTaskDays = userTaskDays.map((day) =>
```
Comment (fshabanov) [inline apps/backend/src/modules/tasks/task.service.ts:25]:
```suggestion
		const normalizedTaskDays = userTaskDays.map((day) =>
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by WilhelmWesser — APPROVED** [review]

File: apps/frontend/src/pages/tasks/libs/components/deadline/deadline.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+			(timeToDeadline % MillisecondsPerUnit.HOUR) / MillisecondsPerUnit.MINUTE,
+		);
+
+		const formattedHours = String(hours).padStart(TimePad.HOURS, TIME_PAD_FILL);
+		const formattedMinutes = String(minutes).padStart(
+			TimePad.MINUTES,
+			TIME_PAD_FILL,
+		);
+
+		setCountdown(`${String(days)}:${formattedHours}:${formattedMinutes}`);
```
Comment (fshabanov) [inline apps/frontend/src/pages/tasks/libs/components/deadline/deadline.tsx:54]: Not sure if displaying days in this format is very clear. Discuss with QA

**Review by fshabanov — APPROVED** [review]
We will merge your PR after [335-feat-add-my-tasks-screen](https://github.com/BinaryStudioAcademy/bsa-2024-bebalance/tree/335-feat-add-my-tasks-screen) is merged

**Comment by fshabanov** [comment]
Fix conflicts and we can merge

File: apps/frontend/src/pages/tasks/libs/components/deadline/deadline.tsx (hunk 3 above)
Comment (luiqor) [reply apps/frontend/src/pages/tasks/libs/components/deadline/deadline.tsx]: I discussed with QA's about countdown format. We got a new format that received the most votes after discussion. I changed the old one to the new.

image

**Review by fshabanov — APPROVED** [review]

**Pull request merged**